File: Src/Icm.Core/AI Search/AaSearch.vb
Imports System.Runtime.CompilerServices

Namespace Icm.Collections.Generic
    Public Module AaSearch

        <Extension()>
        Function f(ByVal n As INode) As Integer
            Return n.g + n.h
        End Function

        Public Function AaSearch(ByVal root As INode) As INode
            Dim visited As New HashSet(Of INode) From {root}
            Dim evaluated As New HashSet(Of INode)

            root.g = 0

            Do Until visited.Count = 0
                Dim current = visited.MinEntity(Function(n) n.f)

                If current.IsGoal Then
                    Return current
                End If
                visited.Remove(current)
                evaluated.Add(current)
                For Each neighbor In current.Neighbors
                    If evaluated.Contains(neighbor) Then
                        Continue For
                    End If
                    Dim tentative_g = current.g + current.Distance(neighbor)

                    If visited.Contains(neighbor) OrElse tentative_g <= neighbor.g Then
                        neighbor.CameFrom = current
                        neighbor.g = tentative_g
                        If Not visited.Contains(neighbor) Then
                            visited.Add(neighbor)
                        End If

                    End If

                Next
            Loop
            Return Nothing
        End Function

    End Module
End Namespace

File: Src/Icm.Core/AI Search/AaSearch.cs

using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Runtime.CompilerServices;

namespace Icm.Collections.Generic
{
	public static class AaSearch
	{

		[Extension()]
		public static int f(INode n)
		{
			return n.g + n.h;
		}

		public static INode AaSearch(INode root)
		{
			HashSet<INode> visited = new HashSet<INode> { root };
			HashSet<INode> evaluated = new HashSet<INode>();

			root.g = 0;

			while (!(visited.Count == 0)) {
				dynamic current = visited.MinEntity(n => n.f);

				if (current.IsGoal) {
					return current;
				}
				visited.Remove(current);
				evaluated.Add(current);
				foreach (void neighbor_loopVariable in current.Neighbors) {
					neighbor = neighbor_loopVariable;
					if (evaluated.Contains(neighbor)) {
						continue;
					}
					dynamic tentative_g = current.g + current.Distance(neighbor);

					if (visited.Contains(neighbor) || tentative_g <= neighbor.g) {
						neighbor.CameFrom = current;
						neighbor.g = tentative_g;
						if (!visited.Contains(neighbor)) {
							visited.Add(neighbor);
						}

					}

				}
			}
			return null;
		}

	}
}

//=======================================================
//Service provided by Telerik (www.telerik.com)
//Conversion powered by NRefactory.
//Twitter: @telerik
//Facebook: facebook.com/telerik
//=======================================================
